File: solutions/c/armstrong-numbers/2/armstrong_numbers.c
#include "armstrong_numbers.h"
#include <stdbool.h>

int ipow(int base, int exp) {
    int result = 1;
    while (exp) {
        if (exp % 2) result *= base;
        base *= base;
        exp /= 2;
    }
    return result;
}

bool is_armstrong_number(int candidate) {
    int original = candidate;
    int total = 0;
    int digit_num = 0;
    int temp = candidate;

    while (temp) {
        temp /= 10;
        digit_num++;
    }
    if (candidate == 0) digit_num = 1;

    temp = candidate;
    while (temp) {
        int digit = temp % 10;
        total += ipow(digit, digit_num);
        temp /= 10;
    }
    return total == original;
}

File: solutions/python/armstrong-numbers/2/armstrong_numbers.py
def len_num(num):
    if num == 0:
        return 1
    count = 0
    while num:
        num //= 10
        count += 1
    return count


def is_armstrong_number(number):
    original = number
    digit_num = len_num(number)
    total = 0

    while number:
        digit = number % 10
        total += digit ** digit_num
        number //= 10
    
    return original == total